File: 2. Linkedlist/Single Linked List/linkedlist.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

    def setNextNode(self,node):
        self.next = node

    def getNextNode(self):
        return self.next

class linkedlist:
    def __init__(self):
        self.head = None

    def addtohead(self,data):
        newhead = Node(data)
        current = self.head
        self.head = newhead
        if(current != None):
            self.head.setNextNode(current)

    def addtotail(self,data):
        tail = self.head
        if(tail == None):
            self.head = Node(data)
        else:
            while(tail.getNextNode() != None):
                tail = tail.getNextNode()
        tail.setNextNode(Node(data))

    def removehead(self):
        removedhead = self.head
        if(removedhead == None):
            return None
        self.head = removedhead.getNextNode()
        return removedhead.data

    def printlist(self):
        output = "<head> "
        currentNode = self.head
        while(currentNode != None):
            output += currentNode.data+ " "
            currentNode = currentNode.getNextNode()
        output += " <tail>"
        print(output)
        return output

seasons = linkedlist()

seasons.printlist()
seasons.addtohead("summer")
seasons.addtohead("spring")
seasons.printlist()
seasons.addtotail("fall")
seasons.addtotail("fall")
seasons.printlist()
seasons.removehead()
seasons.printlist()

File: 2. Linkedlist/Single Linked List/LinkedList.java
public class LinkedList{
  public static void main(String []args){
    LinkedList seasons = new LinkedList();
    //Command
    seasons.printList();
    seasons.addToHead("summer");
    seasons.addToHead("spring");
    seasons.printList();
    seasons.addToTail("fall");
    seasons.addToTail("winter");
    seasons.printList();
    seasons.removehead();
    seasons.printList();
  }

  public Node head;

  public LinkedList(){
    this.head = null;
  }
  public void addToHead(String data){
    Node newhead = new Node(data);
    Node current = this.head;
    this.head = newhead;
    if(current != null) this.head.setNextNode(current);
  }
  public void addToTail(String data){
    Node tail = this.head;
    if(tail == null) this.head = new Node(data);
    else{
      while(tail.getNextNode() != null) tail = tail.getNextNode();
    }
    tail.setNextNode(new Node(data));
  }
  public String removehead(){
    Node removedhead = this.head;
    if(removedhead == null) return null;
    this.head = removedhead.getNextNode();
    return removedhead.data;
  }
  public String printList(){
    String output = "<head> ";
    Node currentNode = this.head;
    while(currentNode != null){
      output += currentNode.data+" ";
      currentNode = currentNode.getNextNode();
    }
    output += " <tail>";
    System.out.println(output);
    return output;
  }
}

// https://www.codecademy.com/learn/paths/pass-the-technical-interview-with-java/tracks/linear-data-structures-java/modules/singly-linked-lists-java/cheatsheet
